### PR TITLE
Added How_JAX_primitives_work colab

### PR DESCRIPTION
**Pull request by gnecula**
This addresses in part issue #116.

### CODE REVIEW

**Comment by necula01** [comment]
@shoyer: Should this go into doc/notebooks or in notebooks/. I am not sure what criterion you use to decide this.

**Comment by necula01** [comment]
This Colab can be viewed at https://colab.research.google.com/drive/16o1tffiB6QlGoDVScW9hYsJn2BsTVrpa

**Comment by shoyer** [comment]
There's no particular reason why we need to put notebooks in the "notebooks" subdirectory. I think the docs build will work fine with them anywhere inside the docs directory.

One thing I do recommend is being careful with your use of top level (`#`) and lower level (`##`, `###`, etc) headings. The style of the docs pages tends to makes these a bit more noticible than colab, so check your table of contents first to make you have the right hierarchy.

**Comment by shoyer** [comment]
Oh. I see now that you were asking about docs/notebooks vs notebooks :).

I don't think there is any official policy, but my inclination would be to move over everything into the docs, whch I find to be much easier to navigate/browse. The only reason why I didn't do everything is that it takes a bit of tweaking to make notebooks look good after migrating, and I hadn't gotten around to it.

**Comment by gnecula** [comment]
What do I need to do to make sure that the Colab looks good (other than aligning the use of # with the other colabs)?

**Comment by shoyer** [comment]
You can run a docs build locally if you want to check it first... or we can just merge this (once we're happy with the content) and check/fix the formatting once it builds automatically after the merge.

File: docs/notebooks/How_JAX_primitives_work.ipynb
Comment (shoyer) [inline docs/notebooks/How_JAX_primitives_work.ipynb]: there's an unfinished sentence here

File: docs/notebooks/How_JAX_primitives_work.ipynb
Comment (shoyer) [inline docs/notebooks/How_JAX_primitives_work.ipynb]: Do we need to worry about broadcasting here? What if the inputs have different numbers of dimensions?

File: docs/notebooks/How_JAX_primitives_work.ipynb
Comment (shoyer) [inline docs/notebooks/How_JAX_primitives_work.ipynb]: This might be a good place to mention that possibly traced variables must always be passed by position directly into `.bind()` vs. non-traced variables that can be passed as keyword arguments. This means that if you want to accept structured arguments, you'll need to flatten them (with `jax.tree_util`) before calling the primitive itself.

File: docs/notebooks/How_JAX_primitives_work.ipynb
Comment (gnecula) [reply docs/notebooks/How_JAX_primitives_work.ipynb]: Agreed, all "ma_" replaced by "multiply_add_" and "sq_add" by "square_add".

File: docs/notebooks/How_JAX_primitives_work.ipynb
Comment (gnecula) [reply docs/notebooks/How_JAX_primitives_work.ipynb]: I added a bit of clarification that this works when the inputs have the same dimensions. Do you think we need to do something more elaborate for the purposes of this doc?

File: docs/notebooks/How_JAX_primitives_work.ipynb
Comment (gnecula) [reply docs/notebooks/How_JAX_primitives_work.ipynb]: Added a clarification.

**Review by shoyer — APPROVED** [review]
Looks good to me!

**Review by froystig — APPROVED** [review]

File: docs/notebooks/How_JAX_primitives_work.ipynb
Comment (jekbradbury) [inline docs/notebooks/How_JAX_primitives_work.ipynb]: I'd add a note to the effect that this interface isn't expected to be stable, and might change with revisions to JAX's core.

File: docs/notebooks/How_JAX_primitives_work.ipynb
Comment (jekbradbury) [inline docs/notebooks/How_JAX_primitives_work.ipynb]:
```suggestion
        "The right way to add support for multiply-add is in terms of existing\n",
```

File: docs/notebooks/How_JAX_primitives_work.ipynb
Comment (jekbradbury) [inline docs/notebooks/How_JAX_primitives_work.ipynb]:
```suggestion
        "We need to tell JAX how to evaluate the batched version of the primitive. In this particular case, the `multiply_add_prim` already operates pointwise for any dimension of input vectors. So the batched version can use the same `multiply_add_prim` implementation."
```

**Review by jekbradbury — APPROVED** [review]

File: docs/notebooks/How_JAX_primitives_work.ipynb
Comment (jekbradbury) [reply docs/notebooks/How_JAX_primitives_work.ipynb]: Oops, I think the "suggested change" interface broke here. This was another typofix, multiple_add_prim -> multiply_add_prim.

File: docs/notebooks/How_JAX_primitives_work.ipynb
Comment (gnecula) [reply docs/notebooks/How_JAX_primitives_work.ipynb]: I would rather not add this note. This colab is executable, so if the interface changes either the code will break (and we will know), or the trace will change.

**Pull request merged**
